components/service/experiments/src/test/java/mozilla/components/service/experiments/ActiveExperimentTest.kt

```diff
@@ -0,0 +1,144 @@
+                )
+            ),
+            testDefaultExperiment(
+                id = "id2",
+                branches = listOf(
+                    Experiment.Branch("branch3", 2)
+                )
+            )
+        )
+
+        // Start with empty storage.
+        assertNull(ActiveExperiment.load(context, experiments))
+        verify(context).getSharedPreferences(prefNameActiveExperiment, Context.MODE_PRIVATE)
+        verify(sharedPrefs).getString(prefKeyId, null)
+
+        // Test valid experiment id and branch name variations.
+        setActiveExperimentPrefs("id1", "branch1")
+        var active = ActiveExperiment.load(context, experiments)
+        assertEquals(experiments[0], active!!.experiment)
+        assertEquals("branch1", active!!.branch)
+
+        setActiveExperimentPrefs("id1", "branch2")
+        active = ActiveExperiment.load(context, experiments)
+        assertEquals(experiments[0], active!!.experiment)
+        assertEquals("branch2", active!!.branch)
+
+        setActiveExperimentPrefs("id2", "branch3")
+        active = ActiveExperiment.load(context, experiments)
+        assertEquals(experiments[1], active!!.experiment)
+        assertEquals("branch3", active!!.branch)
+
+        // Test invalid experiment id and branch name variations.
+        testReset()
+        setActiveExperimentPrefs("id2", "non-existent")
+        assertNull(ActiveExperiment.load(context, experiments))
+        verify(context).getSharedPreferences(prefNameActiveExperiment, Context.MODE_PRIVATE)
+        verify(sharedPrefs).getString(prefKeyId, null)
+        verify(sharedPrefs).getString(prefKeyBranch, null)
+
+        setActiveExperimentPrefs("non-existent", "branch1")
+        assertNull(ActiveExperiment.load(context, experiments))
+        verify(context, times(2)).getSharedPreferences(prefNameActiveExperiment, Context.MODE_PRIVATE)
+        verify(sharedPrefs, times(2)).getString(prefKeyId, null)
+        verify(sharedPrefs, times(2)).getString(prefKeyBranch, null)
+    }
+
+    @Test
+    fun `test saving state`() {
+        testReset()
+
+        val experiment = testDefaultExperiment(
+            id = "some-id",
+            branches = listOf(Experiment.Branch("some-branch", 3))
+        )
+        val active = ActiveExperiment(experiment, "some-branch")
+
+        active.save(context)
+        verify(context).getSharedPreferences(prefNameActiveExperiment, Context.MODE_PRIVATE)
+        verify(sharedPrefs).edit()
+        verify(prefsEditor).putString(prefKeyId, "some-id")
+        verify(prefsEditor).putString(prefKeyBranch, "some-branch")
+        verify(prefsEditor).apply()
+    }
+
+    @Test
+    fun `test clearing state`() {
+        testReset()
+
+        ActiveExperiment.clear(context)
+        verify(context).getSharedPreferences(prefNameActiveExperiment, Context.MODE_PRIVATE)
+        verify(sharedPrefs).edit()
+        verify(prefsEditor).clear()
+        verify(prefsEditor).apply()
+    }
+}
```

Comment:
nit: no blank line at end of file (not a linter rule for some reason but most of our files have a blank line at the end)